galsim/table.py

```diff
@@ -645,76 +666,110 @@ def _call_constant(self, x, y):
         else:
             x = np.array(x, dtype=float, copy=False)
             y = np.array(y, dtype=float, copy=False)
-            f = np.empty_like(x, dtype=float)
-            f.fill(self.constant)
-            good = ((x >= self.x[0]) & (x <= self.x[-1]) &
-                    (y >= self.y[0]) & (y <= self.y[-1]))
-            xx = np.ascontiguousarray(x[good].ravel(), dtype=float)
-            yy = np.ascontiguousarray(y[good].ravel(), dtype=float)
-            tmp = np.empty_like(xx, dtype=float)
-            self._tab.interpMany(xx.ctypes.data, yy.ctypes.data, tmp.ctypes.data, len(xx))
-            f[good] = tmp
-            return f
-
-    def _call_raise(self, x, y):
+            if grid:
+                f = np.empty((len(x), len(y)), dtype=float)
+                # Fill in interpolated values first, then go back and fill in
+                # constants
+                xx = np.ascontiguousarray(x.ravel(), dtype=float)
+                yy = np.ascontiguousarray(y.ravel(), dtype=float)
+                self._tab.interpGrid(xx.ctypes.data, yy.ctypes.data, f.ctypes.data, len(xx), len(yy))
+                badx = (x < self.x[0]) | (x > self.x[-1])
+                bady = (y < self.y[0]) | (y > self.y[-1])
+                f[badx,:] = self.constant
+                f[:, bady] = self.constant
+                return f
+            else:
+                # Start with constant array, then interpolate good positions
+                f = np.empty_like(x, dtype=float)
+                f.fill(self.constant)
+                good = ((x >= self.x[0]) & (x <= self.x[-1]) &
+                        (y >= self.y[0]) & (y <= self.y[-1]))
+                xx = np.ascontiguousarray(x[good].ravel(), dtype=float)
+                yy = np.ascontiguousarray(y[good].ravel(), dtype=float)
+                tmp = np.empty_like(xx, dtype=float)
+                self._tab.interpMany(xx.ctypes.data, yy.ctypes.data, tmp.ctypes.data, len(xx))
+                f[good] = tmp
```

Comment:
I'm curious about the asymmetry of good vs bad implementations here.  I don't think I have a particularly good intuition about which is faster for typical use cases.  Was there some motivation for the difference?  Or just that you coded them up at different times and chose arbitrarily and differently.  :)

Reply:
No good reason.  The bad reason though was that I haven't found a satisfying way to assign to a numpy array indexed with a boolean mask in more than one dimension, i.e., something like

```python
f[goodx, goody] = blah
```

FWIW, I'm not really sure we _have_ a typical use case for `grid=True`, though I think it is a nice option to have.

Reply:
OK.  I thought the phase_psf stuff was on a grid.  No?

Reply:
We might be able to use a grid for a pure optical screen, but since the atmospheric screens depend on time and they're moving around, we generally aren't evaluating those on a grid.